KIPs/kip-146.md

```diff
@@ -0,0 +1,157 @@
+---
+kip: 146
+title: Random Proposer Selection
+author: Ian (@ian0371), Ollie (@blukat29), Joseph (@jiseongnoh), and Aidan (@aidan-kwon)
+discussions-to: https://github.com/klaytn/kips/issues/146
+status: Draft
+type: Standards Track
+category: Core
+created: 2023-06-15
+---
+
+## Simple Summary
+
+An unpredictable and verifiable proposer selection policy
+
+## Abstract
+
+This standard outlines an approach for a random proposer selection policy. A proposer is determined every block in a random manner. The randomness relies on [KIP-114](https://github.com/klaytn/kips/blob/main/KIPs/kip-114.md), thereby rendering the new policy both unpredictable and verifiable.
+
+## Motivation
+
+Klaytn has adopted the Byzantine Fault Tolerance (BFT) consensus mechanism, where a block proposer selected by a deterministic algorithm proposes the next block. A vulnerable point of BFT is that an adversary can effectively halt the entire network by simply targeting the proposer, a.k.a. targeted DoS. What makes Klaytn even more vulnerable to a targeted DoS attack is that the current proposer selection policy enables the prediction of up to `proposerUpdateInterval` (3600 in case of Cypress) number of proposers in advance.
+
+As of now, a targeted DoS is practically infeasible in Klaytn because it is a permissioned chain which allows only authorized validators to join the network. However, in permissionless network, a targeted DoS is a viable option for attackers. A random proposer selection can increase the difficulty of the attack by introducing an uncertainty of the upcoming proposers.
+
+Meanwhile, a few malicious validators must not be able to forge the randomness to arbitrarily determine the upcoming proposers. If not, it may increase the risk of censorship or monopolizing block rewards. Therefore, the new random proposer selection policy should be both unpredictable and verifiable.
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.
+
+### Parameters
+
+| Constant     | Value |
+| ------------ | ----- |
+| `FORK_BLOCK` | TBD   |
+
+### Shuffling
+
+A shuffling algorithm must be defined which will act as a building block of proposer and committee selection.
+
+[Fisher-Yates shuffle](https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle) must be used in the below pseudocode (i.e., `random.shuffle`), and [Golang's RNG](https://cs.opensource.google/go/go/+/master:src/math/rand/rng.go;l=7-12) must be used as a pseudo random number generator internally in Fisher-Yates.
```

Comment:
Is it possible to check the performance of the shuffle function?

Reply:
I'll add a performance of shuffling subsection under rationale, and attach a benchmark.

Reply:
I think the benchmark result is meaningless, 'cause it's extremely fast. I don't think this should be in KIP.

result:
```
BenchmarkShuffleIter/30-8       1000000000               0.0000145 ns/op
BenchmarkShuffleIter/50-8       1000000000               0.0000150 ns/op
BenchmarkShuffleIter/100-8      1000000000               0.0000162 ns/op
BenchmarkShuffleIter/1000-8     1000000000               0.0000547 ns/op
BenchmarkShuffleRecursive/30-8          1000000000               0.0000897 ns/op
BenchmarkShuffleRecursive/50-8          1000000000               0.0001098 ns/op
BenchmarkShuffleRecursive/100-8         1000000000               0.0000761 ns/op
BenchmarkShuffleRecursive/1000-8        1000000000               0.0001135 ns/op
```

benchmark code:
```go
func BenchmarkShuffleIter(b *testing.B) {
	run := func(N int) func(b *testing.B) {
		return func(b *testing.B) {
			validators := genValidators(N)
			b.ResetTimer()
			shuffleIter(validators, 4, 2222)
		}
	}

	numVals := []int{30, 50, 100, 1000}
	for _, N := range numVals {
		b.Run(fmt.Sprintf("%d", N), run(N))
	}
}

func BenchmarkShuffleRecursive(b *testing.B) {
	run := func(N int) func(b *testing.B) {
		return func(b *testing.B) {
			validators := genValidators(N)
			b.ResetTimer()
			shuffleRecursive(validators, 4, 2222)
		}
	}

	numVals := []int{30, 50, 100, 1000}
	for _, N := range numVals {
		b.Run(fmt.Sprintf("%d", N), run(N))
	}
}
```